segmentstore/storage/impl/src/test/java/io/pravega/segmentstore/storage/impl/exts3/EXTS3StorageTest.java

```diff
@@ -0,0 +1,281 @@
+    }
+
+    /**
+     * Wrapper over S3JerseyClient. This implements ACLs, multipart copy and multiple writes to the same object on top of S3Proxy implementation.
+     */
+    public static class S3JerseyClientWrapper extends S3JerseyClient {
+        private static final ConcurrentMap<String, AclSize> ACL_MAP = new ConcurrentHashMap<>();
+
+        public S3JerseyClientWrapper(S3Config exts3Config) {
+            super(exts3Config);
+        }
+
+        @Override
+        public PutObjectResult putObject(PutObjectRequest request) {
+            S3ObjectMetadata metadata = request.getObjectMetadata();
+
+           if ( request.getObjectMetadata() != null ) {
+            request.setObjectMetadata(null);
+           }
+            PutObjectResult retVal = super.putObject(request);
+           if (request.getAcl() != null) {
+               long size = 0;
+               if (request.getRange() != null) {
+                   size = request.getRange().getLast() -1;
+               }
+               ACL_MAP.put(request.getKey(), new AclSize(request.getAcl(),
+                       size));
+           }
+           return retVal;
+        }
+
+        @Override
+        public void putObject(String bucketName, String key, Range range, Object content) {
+            byte[] totalByes = new byte[Math.toIntExact(range.getLast() + 1)];
+            try {
+                if ( range.getFirst() != 0 ) {
+                    int bytesRead = getObject(bucketName, key).getObject().read(totalByes, 0,
+                            Math.toIntExact(range.getFirst()));
+                    if ( bytesRead != range.getFirst() ) {
+                        throw new CompletionException(new BadOffsetException(key, range.getFirst(), bytesRead));
+                    }
+                }
+                int bytesRead = ( (InputStream) content).read(totalByes, Math.toIntExact(range.getFirst()),
+                        Math.toIntExact(range.getLast() + 1 - range.getFirst()));
+
+                if ( bytesRead != range.getLast() + 1 - range.getFirst()) {
+                    throw new IllegalArgumentException();
+                }
+                super.putObject( new PutObjectRequest(bucketName, key, (Object) new ByteArrayInputStream(totalByes)));
+                ACL_MAP.get(key).setSize(range.getLast() -1);
+            } catch (IOException e) {
+            }
+        }
+
+        @Override
+        public void setObjectAcl(String bucketName, String key, AccessControlList acl) {
+            AclSize retVal = ACL_MAP.get(key);
+            if ( retVal == null ) {
+                throw new S3Exception("NoObject", 500, "NoSuchKey", key);
+            }
+            retVal.setAcl(acl);
+        }
+
+        @Override
+        public void setObjectAcl(SetObjectAclRequest request) {
+            AclSize retVal = ACL_MAP.get(request.getKey());
+            if ( retVal == null ) {
+                throw new S3Exception("NoObject", 500, "NoSuchKey", request.getKey());
```

Comment:
HTTP code 500 means Internal Server Error. Please use an appropriate HTTP code for each.

Also, isn't there some class somewhere in java that has these listed as constants?